src/Storages/System/StorageSystemReplicas.cpp

```diff
@@ -151,14 +152,48 @@ Pipe StorageSystemReplicas::read(
 
     MutableColumns res_columns = storage_snapshot->metadata->getSampleBlock().cloneEmptyColumns();
 
-    for (size_t i = 0, size = col_database->size(); i < size; ++i)
+    size_t tables_size = col_database->size();
+    size_t thread_pool_size = std::min(tables_size, static_cast<size_t>(getNumberOfPhysicalCPUCores()));
+
+    auto settings = context->getSettingsRef();
+    if (settings.max_threads != 0)
+        thread_pool_size = std::min(thread_pool_size, static_cast<size_t>(settings.max_threads));
+
+    ThreadPool thread_pool(thread_pool_size);
+    std::atomic<bool> error_flag = false;
+    Exception exception;
+
+    std::vector<StorageReplicatedMergeTree::Status> statuses(tables_size);
+
+    for (size_t i = 0; i < tables_size; ++i)
     {
-        StorageReplicatedMergeTree::Status status;
-        dynamic_cast<StorageReplicatedMergeTree &>(
-            *replicated_tables
-                [(*col_database)[i].safeGet<const String &>()]
-                [(*col_table)[i].safeGet<const String &>()]).getStatus(status, with_zk_fields);
+        thread_pool.scheduleOrThrowOnError([&, i=i]
+        {
+            try
+            {
+                dynamic_cast<StorageReplicatedMergeTree &>(
+                *replicated_tables
+                    [(*col_database)[i].safeGet<const String &>()]
+                    [(*col_table)[i].safeGet<const String &>()]).getStatus(statuses[i], with_zk_fields);
+            }
+            catch (...)
+            {
+                tryLogCurrentException("system.replicas", "Failed to fetch system.replicas data");
 
+                /// We capture one of the exceptions to be thrown later
```

Comment:
It's not a problem that tasks may throw. The problem is that if some task throws then `statuses` will be destroyed before `thread_pool`. Some thread may be still working and may try to access `statuses`.

Reply:
Yes, I now see the problem. Changed it a bit, so now it looks like it should be ok.